File: course/apps/api/app/api/v1/ai_tutor.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.core.deps import get_current_user
from app.core.response import api_response
from app.db.mongodb import get_read_db
from app.services.ai_tutor import ask_ai_tutor, get_chat_history, clear_chat_history

router = APIRouter()


class AskQuestionIn(BaseModel):
    question: str = Field(..., min_length=1, max_length=2000)


@router.post("/courses/{course_id}/lessons/{lesson_id}/ai-tutor/ask")
async def ask_question(
    course_id: str,
    lesson_id: str,
    body: AskQuestionIn,
    user: dict = Depends(get_current_user),
):
    """Ask a question to the AI tutor for a specific lesson."""
    db = get_read_db()

    course = await db.courses.find_one({"_id": course_id})
    if not course:
        raise HTTPException(status_code=404, detail="Course not found")

    lesson = None
    for l in course.get("syllabus", []):
        if l["id"] == lesson_id:
            lesson = l
            break
    if not lesson:
        raise HTTPException(status_code=404, detail="Lesson not found")

    result = await ask_ai_tutor(user["id"], course, lesson, body.question)
    return api_response(result)


@router.get("/courses/{course_id}/lessons/{lesson_id}/ai-tutor/history")
async def get_history(
    course_id: str,
    lesson_id: str,
    user: dict = Depends(get_current_user),
):
    """Get AI tutor conversation history for a lesson."""
    history = await get_chat_history(user["id"], course_id, lesson_id)
    return api_response({"messages": history})


@router.delete("/courses/{course_id}/lessons/{lesson_id}/ai-tutor/history")
async def delete_history(
    course_id: str,
    lesson_id: str,
    user: dict = Depends(get_current_user),
):
    """Clear AI tutor conversation history for a lesson."""
    cleared = await clear_chat_history(user["id"], course_id, lesson_id)
    return api_response({"cleared": cleared})

File: course/apps/api/src/routes/ai_tutor.ts
import { Hono } from "hono";
import { Env, Variables } from "../types";
import { query, queryOne, execute, apiResponse, successResponse, errorResponse, internalErrorResponse } from "../lib/db";

const app = new Hono<{ Bindings: Env; Variables: Variables }>();

function getCourseForLesson(env: Env, lessonId: string): Promise<any> {
  return new Promise(async (resolve) => {
    const courses = await query<any>(env, "SELECT * FROM courses");
    for (const course of courses) {
      if (course.syllabus?.some((l: any) => l.id === lessonId)) {
        resolve(course);
        return;
      }
    }
    resolve(null);
  });
}

app.post("/:course_id/lessons/:lesson_id/ai-tutor/ask", async (c) => {
  try {
    const user = c.get("user");
    if (!user) {
      return c.json({ success: false, data: null, error: "Unauthorized", meta: null }, 401);
    }
    const courseId = c.req.param("course_id");
    const lessonId = c.req.param("lesson_id");
    const body = await c.req.json<{ question: string }>();
    const { question } = body;

    const course = await queryOne<any>(c.env, "SELECT * FROM courses WHERE _id = ?", [courseId]);
    if (!course) {
      return c.json({ success: false, data: null, error: "Course not found", meta: null }, 404);
    }

    const lesson = (course.syllabus || []).find((l: any) => l.id === lessonId);
    if (!lesson) {
      return c.json({ success: false, data: null, error: "Lesson not found", meta: null }, 404);
    }

    const sessionId = `ai-session-${user.sub}-${courseId}-${lessonId}`;
    let session = await queryOne<any>(
      c.env,
      "SELECT * FROM ai_tutor_sessions WHERE _id = ?",
      [sessionId]
    );

    if (!session) {
      const now = new Date().toISOString();
      await execute(
        c.env,
        "INSERT INTO ai_tutor_sessions (_id, user_id, course_id, lesson_id, messages, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
        [sessionId, user.sub, courseId, lessonId, JSON.stringify([]), now, now]
      );
      session = { _id: sessionId, user_id: user.sub, course_id: courseId, lesson_id: lessonId, messages: JSON.stringify([]), created_at: now, updated_at: now };
    }

    const messages = JSON.parse(session.messages || "[]");
    messages.push({ role: "user", content: question, timestamp: new Date().toISOString() });
    const aiResponse = `This is a simulated AI response to: ${question}`;
    messages.push({ role: "assistant", content: aiResponse, timestamp: new Date().toISOString() });

    await execute(
      c.env,
      "UPDATE ai_tutor_sessions SET messages = ?, updated_at = ? WHERE _id = ?",
      [JSON.stringify(messages), new Date().toISOString(), sessionId]
    );

    return c.json(apiResponse(true, { message: aiResponse }, null, null), 200);
  } catch (error) {
    return internalErrorResponse(error);
  }
});

app.get("/:course_id/lessons/:lesson_id/ai-tutor/history", async (c) => {
  try {
    const user = c.get("user");
    if (!user) {
      return c.json({ success: false, data: null, error: "Unauthorized", meta: null }, 401);
    }
    const courseId = c.req.param("course_id");
    const lessonId = c.req.param("lesson_id");

    const sessionId = `ai-session-${user.sub}-${courseId}-${lessonId}`;
    const session = await queryOne<any>(
      c.env,
      "SELECT * FROM ai_tutor_sessions WHERE _id = ?",
      [sessionId]
    );

    const messages = session ? JSON.parse(session.messages || "[]") : [];
    return c.json(apiResponse(true, { messages }, null, null), 200);
  } catch (error) {
    return internalErrorResponse(error);
  }
});

app.delete("/:course_id/lessons/:lesson_id/ai-tutor/history", async (c) => {
  try {
    const user = c.get("user");
    if (!user) {
      return c.json({ success: false, data: null, error: "Unauthorized", meta: null }, 401);
    }
    const courseId = c.req.param("course_id");
    const lessonId = c.req.param("lesson_id");

    const sessionId = `ai-session-${user.sub}-${courseId}-${lessonId}`;
    const session = await queryOne<any>(
      c.env,
      "SELECT * FROM ai_tutor_sessions WHERE _id = ?",
      [sessionId]
    );

    if (session) {
      await execute(
        c.env,
        "UPDATE ai_tutor_sessions SET messages = ?, updated_at = ? WHERE _id = ?",
        [JSON.stringify([]), new Date().toISOString(), sessionId]
      );
    }

    return c.json(apiResponse(true, { cleared: true }, null, null), 200);
  } catch (error) {
    return internalErrorResponse(error);
  }
});

export default app;
